### PR TITLE
Barrier

### PR DESCRIPTION
**Pull request by jackpoz**
This PR includes the work done at https://github.com/bear-on-the-job/arcdps_healing_stats plus some cleanup to reduce branch differences and make the review easier.
Credits go to @bear-on-the-job bear-on-the-job , the original author of the code.
I have been using this for a few weeks and I have had no issues that I could find.

Changes to the original version:
- [x] include/exclude barrier generation per window rather than global
- [x] include/exclude healing per window
- [x] add an option under the "Sort by" option where you can select to sort by barrier generation (ascending/descending) and one for healing + barrier as well ("Total outgoing" maybe?)
- [x] add a new window preset which is "Peers Barrier Generation"
- [x] use "Barrier generation" instead of "Barrier"

Suggestions not included:
- paint the bars in the window based on how much you have of each? So if you do 70% healing and 30% barrier generation, we paint a bar that's 70% green then 30% beige ?

See https://github.com/bear-on-the-job/arcdps_healing_stats/releases/tag/v2.8.1.1 for details about the original version.

### CODE REVIEW

**Comment by jackpoz** [comment]
As you know the code the best,  are these changes something that you could implement,  using this PR as a draft ? Barrier is used a lot in wvw, where your healing addon is a must-have for healers and support classes, so it would make a lot of players happy to have barrier support.

**Comment by jackpoz** [comment]
Moved exclusion per window:
image

image

**Comment by jackpoz** [comment]
About
>paint the bars in the window based on how much you have of each? So if you do 70% healing and 30% barrier generation, we paint a bar that's 70% green then 30% beige ?

I don't think this makes sense. It's a lot easier to see what is healing and what is barrier by keeping the same color for each:
image

Even without reading the title of each window, keeping green for heal and yellow for barrier makes it very easy to recognize what each window is showing.

**Comment by jackpoz** [comment]
As for
>Can we try to consistently call it "barrier generation" rather than just "barrier"
should we call it "barrier gen" in places with little space, like the details window ?
image

**Comment by jackpoz** [comment]
Added preset "Peers barrier generation":
image

By default "Peers outgoing" excludes barrier (unless it should be changed):
image

**Comment by jackpoz** [comment]
Implemented sorting.

Sort by heal only, descending:
image

Sort by barrier only, descending:
image

Sort by total outgoing:
image

**Comment by Krappa322** [comment]
> paint the bars in the window based on how much you have of each? So if you do 70% healing and 30% barrier generation, we paint a bar that's 70% green then 30% beige ?

I think I just badly worded it - I meant what you have here. The first PR colored them all green iirc.

Thanks so much for this, will try to look into it asap!

File: src/GUI.cpp [hunk 1]
```diff
@@ -61,8 +61,20 @@ static void Display_DetailsWindow(HealWindowContext& pContext, DetailsWindowStat
...
+	{
+		ImGui::Text("total healing");
+		ImGuiEx::TextRightAlignedSameLine("%llu", adjustedHealing);
+	}
+
+	if (pState.BarrierGeneration > 0)
+	{
+		ImGui::Text("total barrier gen");
+		ImGuiEx::TextRightAlignedSameLine("%llu", pState.BarrierGeneration);
+	}
```
Comment (Krappa322) [inline src/GUI.cpp:77]: Maybe let's try to make sure at least one of them is always displayed? Something like
```
	if (adjustedHealing > 0 || pState.BarrierGeneration == 0)
	{
		ImGui::Text("total healing");
		ImGuiEx::TextRightAlignedSameLine("%llu", adjustedHealing);
	}

	if (pState.BarrierGeneration > 0)
	{
		ImGui::Text("total barrier gen");
		ImGuiEx::TextRightAlignedSameLine("%llu", pState.BarrierGeneration);
	}
```

File: src/GUI.cpp [hunk 2]
```diff
@@ -73,16 +85,35 @@ static void Display_DetailsWindow(HealWindowContext& pContext, DetailsWindowStat
 		ImGuiEx::TextRightAlignedSameLine("%llu", *pState.Casts);
 	}
 
-	ImGui::Text("healing per second");
-	ImGuiEx::TextRightAlignedSameLine("%.1f", divide_safe(pState.Healing, pState.TimeInCombat));
+	if (adjustedHealing > 0)
```
Comment (Krappa322) [inline src/GUI.cpp:88]: And same thing here

File: src/Options.cpp [hunk 3]
```diff
@@ -56,6 +56,8 @@ bool ReadIni(HealTableOptions& pOptions)
 		pOptions.Windows[i].ExcludeOffSquad = healtable_ini.GetBoolValue(section, "exclude_off_squad", pOptions.Windows[i].ExcludeOffSquad);
 		pOptions.Windows[i].ExcludeMinions = healtable_ini.GetBoolValue(section, "exclude_minions", pOptions.Windows[i].ExcludeMinions);
 		pOptions.Windows[i].ExcludeUnmapped = healtable_ini.GetBoolValue(section, "exclude_unmapped", pOptions.Windows[i].ExcludeUnmapped);
+		pOptions.Windows[i].ExcludeHealing = healtable_ini.GetBoolValue(section, "exclude_healing", pOptions.Windows[i].ExcludeHealing);
+		pOptions.Windows[i].ExcludeBarrierGeneration = healtable_ini.GetBoolValue(section, "exclude_barrier_generation", pOptions.Windows[i].ExcludeBarrierGeneration);
```
Comment (Krappa322) [inline src/Options.cpp:60]: The ini code is just there because of conversion from old format - don't need to add new fields in there

**Review by Krappa322 — CHANGES_REQUESTED** [review]
Looks great overall

**Comment by jackpoz** [comment]
Would you prefer barrier generation to be excluded by default or included by default ? Right now it's excluded by default.

File: src/GUI.cpp (hunk 1 above)
Comment (jackpoz) [reply src/GUI.cpp]: Changed to show both total and per second if there is no data to show:
image

File: src/GUI.cpp (hunk 2 above)
Comment (jackpoz) [reply src/GUI.cpp]: Done, see screenshot above

File: src/Options.cpp (hunk 3 above)
Comment (jackpoz) [reply src/Options.cpp]: Code removed

**Comment by jackpoz** [comment]
Merged master into this branch and fixed the merge conflicts.

**Comment by Krappa322** [comment]
> Would you prefer barrier generation to be excluded by default or included by default ? Right now it's excluded by default.

I think it probably makes the most sense to have it excluded by default. If nothing else because that way it doesn't change all existing windows for users after they upgrade.

**Review by Krappa322 — APPROVED** [review]

**Comment by Krappa322** [comment]
Anything else you'd like to have done? If not I'll merge this and create a pre-release with it

**Comment by jackpoz** [comment]
It's ready for me too, there is plenty of space for more pull requests in the future if we want to change something 😃

**Pull request merged**